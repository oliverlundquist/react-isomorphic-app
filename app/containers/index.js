export { Layout } from './Layout'
export { NotFoundPage } from './NotFoundPage'
export { default as IndexPage } from './IndexPage'
export { default as Users } from './Users'
