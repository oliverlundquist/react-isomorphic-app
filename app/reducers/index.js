export { default as counter } from './CountReducer'
export { default as session } from './SessionReducer'
