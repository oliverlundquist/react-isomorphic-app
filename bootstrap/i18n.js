import { addLocaleData } from 'react-intl'
import en from 'react-intl/locale-data/en'
import ja from 'react-intl/locale-data/ja'
import * as messages from '../lang'

addLocaleData([...en, ...ja]);

export const locales = ['en', 'ja']
export const localeIsValid = locale => locales.indexOf(locale) !== -1
export const getMessages = locale => messages[locale] || {}
