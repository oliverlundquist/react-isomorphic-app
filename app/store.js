import React from 'react';
import { createStore, combineReducers, applyMiddleware } from 'redux'
import { routerReducer } from 'react-router-redux'
import createLogger from 'redux-logger'
import invariant from 'redux-immutable-state-invariant'
import thunk from 'redux-thunk'
import { composeWithDevTools } from 'remote-redux-devtools'
import * as reducers from './reducers'
import { reducer as api } from 'redux-json-api-without-credentials'

export const configureStore = (initialState) => {
	// TODO: move logic to external module and refine validation logic
	const environment = process.env.NODE_ENV || 'development'
	const isClient = () => typeof window !== 'undefined' && environment !== 'testing'
	const isServer = () => ! isClient()

	const enhancer = (environment === 'development' && isClient() === true)
		? composeWithDevTools({ realtime: true, port: 3001 })(applyMiddleware(invariant(), createLogger(), thunk))
		: applyMiddleware(thunk)

	return createStore(
		combineReducers({ routing: routerReducer, api: api, ...reducers }),
		initialState,
		enhancer
	)
}
