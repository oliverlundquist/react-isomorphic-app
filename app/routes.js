import React from 'react'
import { Route, IndexRoute } from 'react-router'
import { Layout, IndexPage, Users, NotFoundPage, Products } from '../app/containers'
import { locales, localeIsValid } from '../bootstrap/i18n'

const checkIfValidLocale = (nextState, replace) => {
	if (! localeIsValid(nextState.params.lang)) {
		replace(locales[0]) // redirect to default language
	}
}
// <Route path=":lang" onEnter={checkIfValidLocale}>

const routes = (
	<Route path="/" component={Layout}>
		<IndexRoute component={IndexPage} />
		<Route path="users" component={Users} />
		<Route path="products" component={Products} />
		{locales.map((locale, i) => (
			<Route path={locale} key={`${locale}_${i}`}>
				<IndexRoute component={IndexPage} />
				<Route path="users" component={Users} />
				<Route path="products" component={Products} />
			</Route>
		))}
		<Route path="*" component={NotFoundPage} />
	</Route>
)

export default routes;
