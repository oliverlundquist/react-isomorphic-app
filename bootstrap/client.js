// import 'babel-polyfill'
import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import { syncHistoryWithStore } from 'react-router-redux'
import { Router, browserHistory } from 'react-router'
import { configureStore } from '../app/store'
import routes from '../app/routes'
import { RadiumProvider } from '../app/providers'
import { IntlProvider, addLocaleData } from 'react-intl'

import en from 'react-intl/locale-data/en';
import ja from 'react-intl/locale-data/ja';
addLocaleData([...en, ...ja]);


const preloadedState = window.__PRELOADED_STATE__
const i18n = window.i18n
const store = configureStore(preloadedState)
const history = syncHistoryWithStore(browserHistory, store);
const rootElement = document.getElementById('app')

// TODO: store jwt token in localstorage if available

render(
	<RadiumProvider>
		<Provider store={store}>
			<IntlProvider locale={i18n.locale} messages={i18n.messages}>
				<Router history={history} onUpdate={() => window.scrollTo(0, 0)} >
					{routes}
				</Router>
			</IntlProvider>
		</Provider>
	</RadiumProvider>,
	rootElement
)
