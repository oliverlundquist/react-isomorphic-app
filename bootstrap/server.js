import React from 'react'
import helmet from 'helmet';
import { renderToString } from 'react-dom/server'
import { match, RouterContext } from 'react-router'
import FaviconMiddleware from './middleware/FaviconMiddleware'
import routes from '../app/routes'
import { configureStore } from '../app/store'
import { Provider } from 'react-redux'

const qs = require('qs');
const express = require('express')
const app = express()
const initialState = {
	counter: { count: 5 }
	// movies: {
	// 	movies: [],
	// 	isFetching: false,
	// 	errorMessage: null,
	// 	firstName: null,
	// 	lastName: null,
	// 	email: null
	// }
};

// expressjs middlewares
app.use(require('response-time')());

// helmet middlewares / security
app.use(helmet.frameguard());
app.use(helmet.xssFilter());
app.use(helmet.noSniff());
app.use(helmet.ieNoOpen());
app.disable('x-powered-by');

app.get('/callback', function (req, res) {
	const parameters    = qs.parse(req.query) // never trust incoming parameters, parseInt them or something
	const grant_type    = 'authorization_code'
	const client_id     = process.env.CLIENT_ID
	const client_secret = process.env.CLIENT_SECRET
	const redirect_uri  = process.env.REDIRECT_URI || 'http://localhost/callback'
	const code          = parameters.code

	res.send('make a post request, store token in localStorage and redirect to index')
})

app.use(FaviconMiddleware)
app.use(express.static('public'))
app.use(handleRender)

function handleRender(req, res) {
	match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
		// error response
		if (error !== null) {
			return res.status(500).send(error.message)
		}
		// redirect response
		if (redirectLocation !== null) {
			return res.redirect(302, redirectLocation.pathname + redirectLocation.search)
		}
		// normal response
		const store = configureStore(initialState)
		const html = renderToString(
			<Provider store={store}>
				<RouterContext {...renderProps} />
			</Provider>
		)
		const finalState = store.getState()
		res.status(200).send(renderFullPage(html, finalState))
	})
}

function renderFullPage(html, preloadedState) {
	return `<!doctype html>
<html>
	<head>
		<title>Redux Universal Example</title>
	</head>
	<body>
		<div id="app">${html}</div>
		<script>
			window.__PRELOADED_STATE__ = ${JSON.stringify(preloadedState).replace(/</g, '\\x3c')}
		</script>
		<script src="/scripts/bundle.js"></script>
	</body>
</html>
`
}

app.listen(3000, function () {
	console.log('Server is listening on port 3000!')
})
