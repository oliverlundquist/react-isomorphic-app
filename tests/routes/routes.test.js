import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { configureStore } from '../../app/store'
import { Router, hashHistory } from 'react-router'
import { syncHistoryWithStore, push } from 'react-router-redux'
import routes from '../../app/routes'
import { IntlProvider } from 'react-intl'
import { locales, getMessages } from '../../bootstrap/i18n'

test('Index route', () => {
	// navigate to uri
	hashHistory.push('/');

	// setup app
	const store = configureStore({ counter: { count: 5 } })
	const history = syncHistoryWithStore(hashHistory, store)
	const component = renderer.create(
		<Provider store={store}>
			<IntlProvider locale={locales[0]} messages={getMessages(locales[0])}>
				<Router history={history} onUpdate={() => window.scrollTo(0, 0)} >
					{routes}
				</Router>
			</IntlProvider>
		</Provider>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('NotFound route', () => {
	// navigate to uri
	hashHistory.push('/some-route-that-doesnt-exist');

	// setup app
	const store = configureStore({ counter: { count: 5 } })
	const history = syncHistoryWithStore(hashHistory, store)
	const component = renderer.create(
		<Provider store={store}>
			<IntlProvider locale={locales[0]} messages={getMessages(locales[0])}>
				<Router history={history} onUpdate={() => window.scrollTo(0, 0)} >
					{routes}
				</Router>
			</IntlProvider>
		</Provider>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Users route', () => {
	// navigate to uri
	hashHistory.push('/users');

	// setup app
	const store = configureStore({ counter: { count: 5 } })
	const history = syncHistoryWithStore(hashHistory, store)
	const component = renderer.create(
		<Provider store={store}>
			<IntlProvider locale={locales[0]} messages={getMessages(locales[0])}>
				<Router history={history} onUpdate={() => window.scrollTo(0, 0)} >
					{routes}
				</Router>
			</IntlProvider>
		</Provider>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
