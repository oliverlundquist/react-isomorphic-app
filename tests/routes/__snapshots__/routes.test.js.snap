exports[`test Index route 1`] = `
<div
  className="app-container">
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "body{margin: 0;
      font-family: Helvetica Neue, Helvetica, Arial, sans-serif;}html{background: #eee;
      font-size: 100%;}@media (min-width: 550px){html{font-size: 120%;}}@media (min-width: 1200px){html{font-size: 140%;}}h1, h2, h3{font-weight: bold;}",
      }
    } />
  <header>
    <a
      href="#/"
      onClick={[Function]}
      style={Object {}}>
      Home
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/users"
      onClick={[Function]}
      style={Object {}}>
      Users
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/products"
      onClick={[Function]}
      style={Object {}}>
      Products
    </a>
    <a
      href="/products">
      Products (direct linkz)
    </a>
  </header>
  <div
    className="app-content">
    <div>
      <span>
        hello
      </span>
      <h1>
        Index Page
      </h1>
      <a
        href="http://passport.dev:8000/oauth/authorize?client_id=7&redirect_uri=http://consumer.dev:8010/callback&response_type=code&scope=read:products read:categories">
        <img
          alt="Login with Mystore.no"
          src="https://s3-eu-west-1.amazonaws.com/mystore-api-auth-app/assets/mystore-api-login-btn.jpg"
          style={
            Object {
              "height": "auto",
              "width": 360,
            }
          } />
      </a>
    </div>
  </div>
  <footer
    style={
      Object {
        "borderTop": "1px solid #000",
        "marginTop": "15px",
      }
    }>
    <p>
      This is where the footer goes.
    </p>
  </footer>
</div>
`;

exports[`test NotFound route 1`] = `
<div
  className="app-container">
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "body{margin: 0;
      font-family: Helvetica Neue, Helvetica, Arial, sans-serif;}html{background: #eee;
      font-size: 100%;}@media (min-width: 550px){html{font-size: 120%;}}@media (min-width: 1200px){html{font-size: 140%;}}h1, h2, h3{font-weight: bold;}",
      }
    } />
  <header>
    <a
      href="#/"
      onClick={[Function]}
      style={Object {}}>
      Home
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/users"
      onClick={[Function]}
      style={Object {}}>
      Users
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/products"
      onClick={[Function]}
      style={Object {}}>
      Products
    </a>
    <a
      href="/products">
      Products (direct linkz)
    </a>
  </header>
  <div
    className="app-content">
    <div>
      <span>
        hello
      </span>
      <h1>
        Index Page
      </h1>
      <a
        href="http://passport.dev:8000/oauth/authorize?client_id=7&redirect_uri=http://consumer.dev:8010/callback&response_type=code&scope=read:products read:categories">
        <img
          alt="Login with Mystore.no"
          src="https://s3-eu-west-1.amazonaws.com/mystore-api-auth-app/assets/mystore-api-login-btn.jpg"
          style={
            Object {
              "height": "auto",
              "width": 360,
            }
          } />
      </a>
    </div>
  </div>
  <footer
    style={
      Object {
        "borderTop": "1px solid #000",
        "marginTop": "15px",
      }
    }>
    <p>
      This is where the footer goes.
    </p>
  </footer>
</div>
`;

exports[`test Users route 1`] = `
<div
  className="app-container">
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "body{margin: 0;
      font-family: Helvetica Neue, Helvetica, Arial, sans-serif;}html{background: #eee;
      font-size: 100%;}@media (min-width: 550px){html{font-size: 120%;}}@media (min-width: 1200px){html{font-size: 140%;}}h1, h2, h3{font-weight: bold;}",
      }
    } />
  <header>
    <a
      href="#/"
      onClick={[Function]}
      style={Object {}}>
      Home
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/users"
      onClick={[Function]}
      style={Object {}}>
      Users
    </a>
    <span
      style={
        Object {
          "margin": "0 5px",
        }
      }>
      |
    </span>
    <a
      href="#/products"
      onClick={[Function]}
      style={Object {}}>
      Products
    </a>
    <a
      href="/products">
      Products (direct linkz)
    </a>
  </header>
  <div
    className="app-content">
    <div>
      <h1>
        Users Page
      </h1>
      <div>
        State inside the Users Container: 
        5
      </div>
      <hr />
      <div>
        Counter Component:
      </div>
      <div
        data-radium={true}>
        5
        <button
          data-radium={true}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#27AE60",
              "borderRadius": "3px",
              "color": "white",
              "fontFamily": "\'Helvetica Neue\', sans-serif",
              "padding": "5px",
              "position": "relative",
              "textAlign": "center",
              "width": "100%",
            }
          }>
          +
        </button>
      </div>
      <div
        data-radium={true}>
        5
        <button
          data-radium={true}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#0074D9",
              "borderRadius": "3px",
              "color": "white",
              "fontFamily": "\'Helvetica Neue\', sans-serif",
              "padding": "5px",
              "position": "relative",
              "textAlign": "center",
              "width": "100%",
            }
          }>
          +
        </button>
      </div>
    </div>
  </div>
  <footer
    style={
      Object {
        "borderTop": "1px solid #000",
        "marginTop": "15px",
      }
    }>
    <p>
      This is where the footer goes.
    </p>
  </footer>
</div>
`;
